### ISSUE TITLE
Outgoing message bubble is misplaced

### ISSUE THREAD

**Issue by nderkach** [jessesquires/JSQMessagesViewController#774]
Using current upstream version (6.1.3) I get the following:

![26388389323297 h1uyi3pg1mfhlefsnyjy_height640](https://cloud.githubusercontent.com/assets/1253785/6095544/24cfb4fa-af94-11e4-8275-4a7df971991a.png)

I've check the code and the first message bubble is indeed initialised with outgoingMessagesBubbleImageWithColor, what gives?

**Comment by mohpor** [comment]
Are you inserting messages to collection view individually (`performBatchUpdates` style) or reload the entire collection upon new messages (default implementation)?
does the problem exists after a refresh (probably a push/pop)?
I've encountered such problems and it was mostly because I'm using my own way of inserting messages to collection view (the current reload-all doesn't work for me) and things get even uglier if you use some indexPath-mapping mechanism (like I do, for the you know, Pagination).
Anyway, I think the problem is in the way your collection view gets reloaded (Wrong cells are getting dequeued)

**Comment by nderkach** [comment]
I'm reloading with ```reloadData``` of a collection view.
As soon, as I add message to an array, I call ```reloadData``` and see the picture above.
It doesn't seem that it's the case of wrong cell dequeued, as the content and order of cells are correct.
It's just seems that for the outgoing cell view it uses ```JSQMessagesCollectionViewCellIncoming.xib``` instead of corresponding outgoing .xib

**Comment by mohpor** [comment]
Are you subclassing `JSQMessage`?
Collection view decides on nib files to load based on the this:
```
    BOOL isOutgoingMessage = [messageSenderId isEqualToString:self.senderId];

```
 property, and it is providing right bubble but wrong cell. These are two different parts of data source . You are doing something right in `-(id<JSQMessageBubbleImageDataSource>)collectionView:(JSQMessagesCollectionView *)collectionView messageBubbleImageDataForItemAtIndexPath:(NSIndexPath *)indexPath` and wrong in  `-(id<JSQMessageData>)collectionView:(JSQMessagesCollectionView *)collectionView messageDataForItemAtIndexPath:(NSIndexPath *)indexPath` .
My guess is `senderID` is not right either in your collection view or your message data.
P.S.: I'm just trying to figure it out, let me know if I'm not helping.

**Comment by nderkach** [comment]
I'm not subclassing.

Here is the relevant code I use to distinguish outgoing and incoming messages:

```
    JSQMessagesBubbleImageFactory *bubbleFactory = [[JSQMessagesBubbleImageFactory alloc] init];

    bubbleImageOutgoing = [bubbleFactory
                                    outgoingMessagesBubbleImageWithColor:kTraffleMainColor];
    
    bubbleImageIncoming = [bubbleFactory
                                    incomingMessagesBubbleImageWithColor:[UIColor whiteColor]];

- (id<JSQMessageBubbleImageDataSource>)collectionView:(JSQMessagesCollectionView *)collectionView messageBubbleImageDataForItemAtIndexPath:(NSIndexPath *)indexPath
{
    JSQMessage *message = messages[indexPath.item];
    
    if ([message.senderId isEqualToString:[PFUser currentUser].objectId])
    {
        return bubbleImageOutgoing;
    }
    return bubbleImageIncoming;
}

- (id<JSQMessageData>)collectionView:(JSQMessagesCollectionView *)collectionView messageDataForItemAtIndexPath:(NSIndexPath *)indexPath
{
    return [messages objectAtIndex:indexPath.item];
}
```

My first guess was the mismatch of ```senderId``` but it doesn't seem to be the case.
What do you mean by "senderID is not right either in your collection view"?

**Comment by mohpor** [comment]
check this out again:

```Objective-C
 BOOL isOutgoingMessage = [messageSenderId isEqualToString:self.senderId];
```
in this snippet, `self.senderId` is the value you send as your **SenderID** in your `JSQMessagesViewController` subclass, right? otherwise the check is always NO and you always get incoming message.

**Comment by nderkach** [comment]
That's it! ```self.senderId``` was set to default here.
Can't imagine, what's the reason to have a default value here, seems like an assert would be more reasonable.

**Comment by mohpor** [comment]
that's right, but to be fair, there is a hint in Getting Started section about it.
Good luck

**Comment by jessesquires** [comment]
Thanks for helping @mohpor !

I agree — I should remove the default value for `self.senderId`.

If I remember correctly, there were issues with this (because of the initialization flow, the assert would always throw), but I'll look into it again.

**Comment by mohpor** [comment]
@jessesquires, Glad I could help.
probably an even better solution would be to add a **Required** method to data source to retrieve setup info for collection view, that would help us avoid race condition and to make sure everybody provides the collection view with minimum required info.

**Comment by jessesquires** [comment]
Fixed in #1636
